docs/Fleet management Design.md

```diff
@@ -101,52 +125,122 @@ type ResourceList map[ResourceName]resource.Quantity
+### Join a member cluster to the fleet
 The managedCluster will look for a secret named `name of the cluster`-`kubeconfig`. 
 For example, the controller that reconciles a `managedCluster` called "member-cluster-A" will look for
-a secret named `member-cluster-A-kubeconfig`. The secret contains the entire kubeconfig data. The controller will then 
-create a namespace of the same name if it does not already exist.
+a secret named `member-cluster-A-kubeconfig`. Initially, the secret can simply contain the entire kubeconfig data.
+
+The controller keeps an in-memory map that keys on the cluster name to track all the information belong to the cluster. 
+One of the information can be the kubeclient created from the secret. Here are a list of operations a controller will 
+perform in the member cluster joining process
+
+1. The controller "claims" the member cluster by creating a "lease" configMap in the member cluster.
+   * The configMap's name is a well known name by convention.
+   * It contains the hub cluster's [unique ID](https://github.com/kubernetes/kubernetes/issues/77487) which is the UID of the namespace kube-system for now.
+   * The controller will reject the member cluster if the configMap already contains a cluster ID not the same as its ID.
+   * We could use a CR but that means we also need to apply a CRD there.
+2. The controller creates a namespace of the same name as the member cluster in the hub cluster.
+3. The controller starts a health check loop for each member cluster which
+   * Checks the health of the member cluster by updating the last health check time in the lease configMap
+   * Lists all the nodes in a member cluster and aggregates their `Capacity` and `Allocatable`
+   * Updating the `managedCluster` condition status
+
+### API definition change proposal
+There is actually a better way to provide the credentials of member clusters to the hub cluster than using a naming convention.
+There is a `ClientConfig` field in the `managedCluster` definition, described above, that is designed to contain the member cluster access info.
+However, the upstream OCM project currently is not utilizing it so that it does not contain the necessary fields to save
+the user credentials. We will propose to add a path to a file that contains the user credentials. The file will be
+on volume mounted to the secret described above. 
+
+Cert rotation is also another concern in the real production environment. In order to support member cluster api-server
+cert rotation, we will have to make the CABundle a list so that it can contain both the new and old server CA certificate.
+
+Here is the new ClientConfig we will propose
+
+```golang
+type ClientConfig struct {
+    // Server URL
+	// +required
+    URL string `json:"url"`
+
+    // CABundle is a list of ca bundle to connect to apiserver of the managed cluster.
+    // System certs are used if it is not set.
+    // +optional
+    CABundles [][]byte `json:"caBundles,omitempty"`
```

Comment:
I wonder how the reconciler knows that there are multiple CA in this bundle? Is there a known format?

Reply:
yes, in the reconcile, it can parse the existing CA, append new and remove old. Something like this https://github.com/open-cluster-management-io/addon-framework/blob/main/pkg/certrotation/cabundle.go#L69